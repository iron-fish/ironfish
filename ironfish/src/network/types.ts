/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import type { ErrorEvent as WSErrorEvent } from 'ws'
import { WebSocketClient } from './webSocketClient'

export enum NetworkMessageType {
  // Core protocol messages
  Identify = 0,
  Disconnecting = 1,
  SignalRequest = 2,
  Signal = 3,
  CannotSatisfyRequest = 4,
  // Peer discovery-related messages
  PeerListRequest = 20,
  PeerList = 21,
  // Syncing-related messages
  // Block syncing
  GetBlockHeadersRequest = 40,
  GetBlockHeadersResponse = 41,
  GetBlocksRequest = 42,
  GetBlocksResponse = 43,
  // Block gossip
  GetCompactBlockRequest = 60,
  GetCompactBlockResponse = 61,
  GetBlockTransactionsRequest = 62,
  GetBlockTransactionsResponse = 63,
  NewBlockHashes = 64,
  NewCompactBlock = 65,
  // Transaction gossip
  PooledTransactionsRequest = 80,
  PooledTransactionsResponse = 81,
  NewPooledTransactionHashes = 82,
  NewTransactions = 83,
}

export type IsomorphicWebSocketConstructor = typeof WebSocket | typeof WebSocketClient
export type IsomorphicWebSocket = WebSocket | WebSocketClient
export type IsomorphicWebSocketErrorEvent = WSErrorEvent
