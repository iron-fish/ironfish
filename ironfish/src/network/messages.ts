/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { SerializedBlock } from '../primitives/block'
import { SerializedTransaction, Transaction } from '../primitives/transaction'
import { IJSON } from '../serde'
import { UnwrapPromise } from '../utils'
import { Identity, isIdentity } from './identity'
import { Gossip } from './messageRouters'

/**
 * The type of the message for the purposes of routing within our code.
 * This includes messages consumed by our connection and peer manager layer,
 * such as identity, signal, and peerList,
 * and message routing types such as gossip, directRPC, and globalRPC.
 */
export enum InternalMessageType {
  identity = 'identity',
  signal = 'signal',
  signalRequest = 'signalRequest',
  peerList = 'peerList',
  peerListRequest = 'peerListRequest',
  cannotSatisfyRequest = 'cannotSatisfyRequest',
  disconnecting = 'disconnecting',
}

export type MessageType = InternalMessageType | string
export type PayloadType = Record<string, unknown> | undefined
/**
 * Used for functions that don't care about the contents of the message.
 */
export type LooseMessage = Message<MessageType, Record<string, unknown>> | Message<MessageType>

/**
 * A message that has been received on the connection. Note that most messages
 * will have other properties, but so long as an object is jsonable and has a
 * type, it's ready to send.
 */
export type Message<
  T extends MessageType,
  P extends PayloadType = undefined,
> = P extends undefined ? { type: T } : { type: T; payload: P }

export type MessagePayload<M> = M extends Message<infer _T, infer P> ? P : never

export function isMessage(obj: unknown): obj is Message<MessageType, PayloadType> {
  if (typeof obj !== 'object' || obj === null) {
    return false
  }
  if (
    'payload' in obj &&
    (typeof (obj as Message<MessageType, Record<string, unknown>>).payload !== 'object' ||
      obj === null)
  ) {
    return false
  }
  return typeof (obj as Message<MessageType, PayloadType>).type === 'string'
}

export function isPayloadMessage(
  obj: unknown,
): obj is Message<MessageType, Record<string, unknown>> {
  return (
    isMessage(obj) &&
    'payload' in obj &&
    typeof obj.payload === 'object' &&
    obj.payload !== null
  )
}

/**
 * Parse a message and verify that it has a type field (passes isMessage)
 *
 * Throws an error if it's not a valid message
 */
export function parseMessage(data: string): Message<MessageType, PayloadType> {
  const message = IJSON.parse(data)
  if (!isMessage(message)) {
    throw new Error('Message must have a type field')
  }
  return message
}

/**
 * A message by which a peer can identify itself to another.
 */
export type Identify = Message<
  InternalMessageType.identity,
  {
    identity: Identity
    name?: string
    version: number
    agent: string
    port: number | null
    head: string
    work: string
    sequence: number
  }
>

export function isIdentify(obj: unknown): obj is Identify {
  if (!isPayloadMessage(obj)) {
    return false
  }

  const payload = obj.payload as Identify['payload']

  return (
    obj.type === InternalMessageType.identity &&
    typeof payload === 'object' &&
    payload !== null &&
    typeof payload.identity === 'string' &&
    typeof payload.agent === 'string' &&
    typeof payload.version === 'number' &&
    typeof payload.head === 'string' &&
    typeof payload.work === 'string' &&
    typeof payload.sequence === 'number'
  )
}

/**
 * A message used to indicate to a peer that we want them to
 * initiatie signaling with us. This is most often used when
 * we discover a peer through another peer but need to indicate
 * to them through a brokering peer to connect to us via webrtc.
 */
export type SignalRequest = Message<
  InternalMessageType.signalRequest,
  {
    sourceIdentity: Identity
    destinationIdentity: Identity
  }
>

export function isSignalRequest(obj: unknown): obj is SignalRequest {
  if (!isPayloadMessage(obj)) {
    return false
  }

  const payload = obj.payload as Signal['payload']
  return (
    obj.type === InternalMessageType.signalRequest &&
    payload !== null &&
    typeof payload.sourceIdentity === 'string' &&
    typeof payload.destinationIdentity === 'string'
  )
}

/**
 * A message used to signal an rtc session between two peers.
 *
 * The referring peer will forward the message to the sourceIdentity,
 * which will need to respond with a signal that has peer and source
 * inverted.
 */
export type Signal = Message<
  InternalMessageType.signal,
  {
    sourceIdentity: Identity
    destinationIdentity: Identity
    nonce: string
    signal: string
  }
>

export function isSignal(obj: unknown): obj is Signal {
  if (!isPayloadMessage(obj)) {
    return false
  }
  const payload = obj.payload as Signal['payload']
  return (
    obj.type === InternalMessageType.signal &&
    payload !== null &&
    typeof payload.sourceIdentity === 'string' &&
    typeof payload.destinationIdentity === 'string' &&
    typeof payload.nonce === 'string' &&
    typeof payload.signal === 'string'
  )
}

/**
 * A message used to request a peer list from another peer.
 *
 * The referring peer will respond with a PeerList message,
 * which contains information about their connected peers.
 */
export type PeerListRequest = Message<InternalMessageType.peerListRequest>

export function isPeerListRequest(obj: unknown): obj is PeerListRequest {
  return isMessage(obj) && obj.type === InternalMessageType.peerListRequest
}

export type PeerList = Message<
  InternalMessageType.peerList,
  {
    connectedPeers: {
      identity: Identity
      name?: string
      address: string | null
      port: number | null
    }[]
  }
>

export function isPeerList(obj: unknown): obj is PeerList {
  if (!isPayloadMessage(obj)) {
    return false
  }
  const payload = obj.payload as PeerList['payload']
  return (
    obj.type === InternalMessageType.peerList &&
    payload !== null &&
    Array.isArray(payload.connectedPeers) &&
    payload.connectedPeers.every((v) => isIdentity(v.identity))
  )
}

export enum DisconnectingReason {
  ShuttingDown = 0,
  Congested = 1,
}

export type DisconnectingMessage = Message<
  InternalMessageType.disconnecting,
  {
    sourceIdentity: Identity
    // Can be null if we're sending the message to an unidentified Peer
    destinationIdentity: Identity | null
    reason: DisconnectingReason
    disconnectUntil: number
  }
>

export function isDisconnectingMessage(obj: unknown): obj is DisconnectingMessage {
  if (!isPayloadMessage(obj)) {
    return false
  }
  const payload = obj.payload as DisconnectingMessage['payload']
  return (
    obj.type === InternalMessageType.disconnecting &&
    payload !== null &&
    typeof payload.sourceIdentity === 'string' &&
    (typeof payload.destinationIdentity === 'string' || payload.destinationIdentity === null) &&
    typeof payload.reason === 'number' &&
    typeof payload.disconnectUntil === 'number'
  )
}

/**
 * A message that we have received from a peer, identified by that peer's
 * identity.
 */
export interface IncomingPeerMessage<M extends Message<MessageType, PayloadType>> {
  peerIdentity: Identity
  message: M
}

/**
 * The type of a Iron Fish message. This is an exhaustive list of
 * the messages that are sent from IronfishNode. Other messages may
 * be sent by peerNetwork's internal mechanisms (for example, a peer list).
 *
 * Note: A Response to a Request must have the same MessageType
 */
export enum NodeMessageType {
  NewBlock = 'NewBlock',
  NewTransaction = 'NewTransaction',
  GetBlockHashes = 'GetBlockHashes',
  GetBlocks = 'GetBlocks',
}

export type GetBlockHashesRequest = Message<
  NodeMessageType.GetBlockHashes,
  {
    start: string | number
    limit: number
  }
>

export type GetBlockHashesResponse = Message<
  NodeMessageType.GetBlockHashes,
  {
    blocks: string[]
  }
>

export type GetBlocksRequest = Message<
  NodeMessageType.GetBlocks,
  {
    start: string | number
    limit: number
  }
>

export type GetBlocksResponse = Message<
  NodeMessageType.GetBlocks,
  {
    blocks: SerializedBlock[]
  }
>

export function isGetBlocksResponse(obj: LooseMessage): obj is GetBlocksResponse {
  if (
    obj.type === NodeMessageType.GetBlocks &&
    'payload' in obj &&
    'blocks' in obj.payload &&
    Array.isArray(obj.payload.blocks)
  ) {
    for (const block of obj.payload.blocks) {
      if (!isBlock(block)) {
        return false
      }
    }

    return true
  }

  return false
}

export function isGetBlocksRequest(obj: PayloadType): obj is GetBlocksRequest['payload'] {
  return (
    obj !== undefined &&
    'start' in obj &&
    (typeof obj.start === 'string' || typeof obj.start === 'number') &&
    'limit' in obj &&
    typeof obj.limit === 'number'
  )
}

export function isGetBlockHashesResponse(obj: LooseMessage): obj is GetBlockHashesResponse {
  if (
    obj.type === NodeMessageType.GetBlockHashes &&
    'payload' in obj &&
    'blocks' in obj.payload &&
    Array.isArray(obj.payload.blocks)
  ) {
    for (const block of obj.payload.blocks) {
      if (!isBlockHash(block)) {
        return false
      }
    }

    return true
  }

  return false
}
export function isGetBlockHashesRequest(
  obj: PayloadType,
): obj is GetBlockHashesRequest['payload'] {
  return (
    obj !== undefined &&
    'start' in obj &&
    (typeof obj.start === 'string' || typeof obj.start === 'number') &&
    'limit' in obj &&
    typeof obj.limit === 'number'
  )
}

function isBlockHash(obj: unknown | undefined): obj is string {
  return typeof obj === 'string'
}

function isBlock(obj: Record<string, unknown> | undefined): obj is SerializedBlock {
  return (
    obj !== undefined &&
    'header' in obj &&
    typeof obj.header === 'object' &&
    obj.header !== null &&
    'hash' in obj.header
  )
}

/**
 * A newly mined block gossipped on the P2P network
 */
export type NewBlock = Message<'NewBlock', { block: SerializedBlock }>

/**
 * Type narrowing to confirm the message payload contains a `block` object.
 * Does not try to confirm whether it is a correct block.
 */
export function isNewBlockPayload(obj: PayloadType): obj is NewBlock['payload'] {
  return (
    obj !== undefined && 'block' in obj && typeof obj.block === 'object' && obj.block !== null
  )
}

/**
 * A newly spent transaction that a client would like to have mined
 */
export type NewTransaction = Message<
  'NewTransaction',
  {
    transaction: SerializedTransaction
  }
>

/**
 * Type narrowing to confirm the message payload contains a `transaction`
 * object. Does not try to validate the transaction.
 */
export function isNewTransactionPayload(obj: PayloadType): obj is NewTransaction['payload'] {
  return (
    obj !== undefined &&
    'transaction' in obj &&
    typeof obj.transaction === 'object' &&
    obj.transaction !== null
  )
}

export type NewBlockMessage = Gossip<
  NodeMessageType.NewBlock,
  UnwrapPromise<{ block: SerializedBlock }>
>

export type NewTransactionMessage = Gossip<
  NodeMessageType.NewTransaction,
  UnwrapPromise<{ transaction: Transaction }>
>
