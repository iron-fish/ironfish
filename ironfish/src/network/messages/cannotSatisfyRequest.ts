/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { NetworkMessageType } from '../types'
import { Direction, RpcNetworkMessage } from './rpcNetworkMessage'

export class CannotSatisfyRequest extends RpcNetworkMessage {
  constructor(rpcId: number) {
    super(NetworkMessageType.CannotSatisfyRequest, Direction.Response, rpcId)
  }

  serializePayload(): void {
    return
  }

  static deserializePayload(rpcId: number): CannotSatisfyRequest {
    return new CannotSatisfyRequest(rpcId)
  }

  getSize(): number {
    return 0
  }
}
