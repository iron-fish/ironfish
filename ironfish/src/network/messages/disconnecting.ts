/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import bufio from 'bufio'
import { Identity, identityLength } from '../identity'
import { NetworkMessageType } from '../types'
import { NetworkMessage } from './networkMessage'

export enum DisconnectingReason {
  ShuttingDown = 0,
  Congested = 1,
  BadMessages = 2,
}

interface CreateDisconnectingMessageOptions {
  // Can be null if we're sending the message to an unidentified Peer
  destinationIdentity: Identity | null
  // Expects a timestamp with millisecond precision
  disconnectUntil: number
  reason: DisconnectingReason
  sourceIdentity: Identity
}

export class DisconnectingMessage extends NetworkMessage {
  readonly destinationIdentity: Identity | null
  readonly disconnectUntil: number
  readonly reason: DisconnectingReason
  readonly sourceIdentity: Identity

  constructor({
    destinationIdentity,
    disconnectUntil,
    reason,
    sourceIdentity,
  }: CreateDisconnectingMessageOptions) {
    super(NetworkMessageType.Disconnecting)
    this.destinationIdentity = destinationIdentity
    this.disconnectUntil = disconnectUntil
    this.reason = reason
    this.sourceIdentity = sourceIdentity
  }

  serializePayload(bw: bufio.StaticWriter | bufio.BufferWriter): void {
    // Truncates the timestamp to seconds
    bw.writeU32(Math.ceil(this.disconnectUntil / 1000))
    bw.writeU8(this.reason)
    bw.writeBytes(Buffer.from(this.sourceIdentity, 'base64'))
    if (this.destinationIdentity) {
      bw.writeBytes(Buffer.from(this.destinationIdentity, 'base64'))
    }
  }

  static deserializePayload(buffer: Buffer): DisconnectingMessage {
    const reader = bufio.read(buffer, true)
    const disconnectUntil = reader.readU32() * 1000
    const reason = reader.readU8()
    const sourceIdentity = reader.readBytes(identityLength).toString('base64')
    let destinationIdentity = null
    if (reader.left()) {
      destinationIdentity = reader.readBytes(identityLength).toString('base64')
    }
    return new DisconnectingMessage({
      destinationIdentity,
      disconnectUntil,
      reason,
      sourceIdentity,
    })
  }

  getSize(): number {
    let size = 0
    size += 4 // disconnectUntil
    size += 1 // reason
    size += identityLength
    if (this.destinationIdentity) {
      size += identityLength
    }
    return size
  }
}
