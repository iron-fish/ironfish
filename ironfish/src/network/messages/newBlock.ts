/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import bufio from 'bufio'
import { SerializedBlock } from '../../primitives/block'
import { NetworkMessageType } from '../types'
import { getBlockSize, readBlock, writeBlock } from '../utils/block'
import { GossipNetworkMessage } from './gossipNetworkMessage'

export class NewBlockMessage extends GossipNetworkMessage {
  readonly block: SerializedBlock

  constructor(block: SerializedBlock, nonce?: Buffer) {
    super(NetworkMessageType.NewBlock, nonce)
    this.block = block
  }

  serialize(): Buffer {
    const bw = bufio.write(this.getSize())

    writeBlock(bw, this.block)

    return bw.render()
  }

  static deserialize(buffer: Buffer, nonce: Buffer): NewBlockMessage {
    const reader = bufio.read(buffer, true)
    const block = readBlock(reader)

    return new NewBlockMessage(block, nonce)
  }

  getSize(): number {
    return getBlockSize(this.block)
  }
}
