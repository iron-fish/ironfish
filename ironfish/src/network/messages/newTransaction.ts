/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import bufio from 'bufio'
import { SerializedTransaction } from '../../primitives/transaction'
import { NetworkMessageType } from '../types'
import { GossipNetworkMessage } from './gossipNetworkMessage'

export class NewTransactionMessage extends GossipNetworkMessage {
  readonly transaction: SerializedTransaction

  constructor(transaction: SerializedTransaction, nonce?: Buffer) {
    super(NetworkMessageType.NewTransaction, nonce)
    this.transaction = transaction
  }

  serialize(): Buffer {
    const bw = bufio.write(this.getSize())
    bw.writeVarBytes(this.transaction)
    return bw.render()
  }

  static deserialize(buffer: Buffer, nonce: Buffer): NewTransactionMessage {
    const reader = bufio.read(buffer, true)
    const transaction = reader.readVarBytes()
    return new NewTransactionMessage(transaction, nonce)
  }

  getSize(): number {
    return bufio.sizeVarBytes(this.transaction)
  }
}
