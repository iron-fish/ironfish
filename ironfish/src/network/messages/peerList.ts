/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import bufio from 'bufio'
import { Assert } from '../../assert'
import { identityLength } from '../identity'
import { NetworkMessageType } from '../types'
import { NetworkMessage } from './networkMessage'

export interface Peer {
  identity: Buffer
  name?: string
  address: string | null
  port: number | null
}

export class PeerListMessage extends NetworkMessage {
  readonly connectedPeers: Peer[]

  constructor(connectedPeers: Peer[]) {
    super(NetworkMessageType.PeerList)
    this.connectedPeers = connectedPeers
  }

  serializePayload(bw: bufio.StaticWriter | bufio.BufferWriter): void {
    bw.writeU16(this.connectedPeers.length)

    for (const peer of this.connectedPeers) {
      const { identity, name, address, port } = peer

      Assert.isEqual(identity.byteLength, identityLength)
      bw.writeBytes(identity)

      let flags = 0
      flags |= Number(!!name) << 0
      flags |= Number(!!port) << 1
      flags |= Number(!!address) << 2
      bw.writeU8(flags)

      if (name) {
        bw.writeVarString(name, 'utf8')
      }

      if (port) {
        bw.writeU16(port)
      }

      if (address) {
        bw.writeVarString(address, 'utf8')
      }
    }
  }

  static deserializePayload(buffer: Buffer): PeerListMessage {
    const reader = bufio.read(buffer, true)
    const connectedPeersLength = reader.readU16()
    const connectedPeers = []

    for (let i = 0; i < connectedPeersLength; i++) {
      const identity = reader.readBytes(identityLength)

      const flags = reader.readU8()
      const hasName = flags & (1 << 0)
      const hasPort = flags & (1 << 1)
      const hasAddress = flags & (1 << 2)

      let name = undefined
      if (hasName) {
        name = reader.readVarString('utf8')
      }

      let port = null
      if (hasPort) {
        port = reader.readU16()
      }

      let address = null
      if (hasAddress) {
        address = reader.readVarString('utf8')
      }

      connectedPeers.push({
        identity,
        name,
        address,
        port,
      })
    }
    return new PeerListMessage(connectedPeers)
  }

  getSize(): number {
    let size = 2

    for (const { name, address, port } of this.connectedPeers) {
      size += identityLength

      size += 1
      if (name) {
        size += bufio.sizeVarString(name, 'utf8')
      }

      if (port) {
        size += 2
      }

      if (address) {
        size += bufio.sizeVarString(address, 'utf8')
      }
    }
    return size
  }
}
