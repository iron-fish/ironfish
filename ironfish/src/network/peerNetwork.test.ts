/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

jest.mock('ws')

import type WSWebSocket from 'ws'
import http from 'http'
import net from 'net'
import ws from 'ws'
import { Assert } from '../assert'
import { mockChain, mockNode, mockStrategy } from '../testUtilities/mocks'
import { DisconnectingMessage } from './messages/disconnecting'
import { NewBlockMessage } from './messages/newBlock'
import { NewTransactionMessage } from './messages/newTransaction'
import { PeerListMessage } from './messages/peerList'
import { PeerNetwork } from './peerNetwork'
import { getConnectedPeer, mockHostsStore, mockPrivateIdentity } from './testUtilities'
import { NetworkMessageType } from './types'

jest.useFakeTimers()

describe('PeerNetwork', () => {
  describe('stop', () => {
    it('stops the peer manager', async () => {
      const peerNetwork = new PeerNetwork({
        identity: mockPrivateIdentity('local'),
        agent: 'sdk/1/cli',
        webSocket: ws,
        node: mockNode(),
        chain: mockChain(),
        strategy: mockStrategy(),
        hostsStore: mockHostsStore(),
      })

      const stopSpy = jest.spyOn(peerNetwork.peerManager, 'stop')
      await peerNetwork.stop()
      expect(stopSpy).toBeCalled()
    })
  })

  describe('when validation fails', () => {
    it('throws an exception', async () => {
      const peerNetwork = new PeerNetwork({
        identity: mockPrivateIdentity('local'),
        agent: 'sdk/1/cli',
        webSocket: ws,
        node: mockNode(),
        chain: mockChain(),
        strategy: mockStrategy(),
        hostsStore: mockHostsStore(),
      })

      const { peer } = getConnectedPeer(peerNetwork.peerManager)
      const message = new PeerListMessage([])
      await expect(
        peerNetwork['handleMessage'](peer, {
          peerIdentity: peer.getIdentityOrThrow(),
          message,
        }),
      ).rejects.not.toBeUndefined()
      await peerNetwork.stop()
    })
  })

  describe('when peers connect', () => {
    it('changes isReady', async () => {
      const peerNetwork = new PeerNetwork({
        identity: mockPrivateIdentity('local'),
        agent: 'sdk/1/cli',
        webSocket: ws,
        node: mockNode(),
        chain: mockChain(),
        strategy: mockStrategy(),
        minPeers: 1,
        hostsStore: mockHostsStore(),
      })

      expect(peerNetwork.isReady).toBe(false)

      const readyChanged = jest.fn()
      peerNetwork.onIsReadyChanged.on(readyChanged)

      peerNetwork.start()
      expect(peerNetwork.isReady).toBe(false)

      const { peer } = getConnectedPeer(peerNetwork.peerManager)
      expect(peerNetwork.isReady).toBe(true)

      peer.close()
      expect(peerNetwork.isReady).toBe(false)

      await peerNetwork.stop()
      expect(peerNetwork.isReady).toBe(false)

      expect(readyChanged).toBeCalledTimes(2)
      expect(readyChanged).toHaveBeenNthCalledWith(1, true)
      expect(readyChanged).toHaveBeenNthCalledWith(2, false)
    })
  })

  describe('when at max peers', () => {
    it('rejects websocket connections', async () => {
      const wsActual = jest.requireActual<typeof WSWebSocket>('ws')

      const peerNetwork = new PeerNetwork({
        identity: mockPrivateIdentity('local'),
        agent: 'sdk/1/cli',
        webSocket: wsActual,
        node: mockNode(),
        chain: mockChain(),
        strategy: mockStrategy(),
        listen: true,
        port: 0,
        minPeers: 1,
        maxPeers: 0,
        hostsStore: mockHostsStore(),
      })

      const rejectSpy = jest
        .spyOn(peerNetwork.peerManager, 'shouldRejectDisconnectedPeers')
        .mockReturnValue(true)

      // Start the network so it creates the webSocketServer
      peerNetwork.start()
      const server = peerNetwork['webSocketServer']
      Assert.isNotUndefined(server, `server`)

      const socket = new net.Socket()
      const req = new http.IncomingMessage(socket)
      const conn = new ws('test')

      const sendSpy = jest.spyOn(conn, 'send').mockReturnValue(undefined)
      const closeSpy = jest.spyOn(conn, 'close').mockReturnValue(undefined)
      server.server.emit('connection', conn, req)
      await peerNetwork.stop()

      expect(rejectSpy).toHaveBeenCalled()
      expect(sendSpy).toHaveBeenCalled()
      expect(closeSpy).toHaveBeenCalled()

      // Check that the disconnect message was serialized properly
      const args = sendSpy.mock.calls[0][0]
      expect(typeof args).toEqual('string')
      const message = JSON.parse(args) as DisconnectingMessage
      expect(message.type).toEqual(NetworkMessageType.Disconnecting)
    })
  })

  describe('when enable syncing is true', () => {
    it('adds new blocks', async () => {
      const peerNetwork = new PeerNetwork({
        identity: mockPrivateIdentity('local'),
        agent: 'sdk/1/cli',
        webSocket: ws,
        node: mockNode(),
        chain: mockChain(),
        strategy: mockStrategy(),
        hostsStore: mockHostsStore(),
      })

      const { peer } = getConnectedPeer(peerNetwork.peerManager)

      const block = {
        header: {
          graffiti: 'chipotle',
          minersFee: '0',
          noteCommitment: {
            commitment: Buffer.from('commitment'),
            size: 1,
          },
          nullifierCommitment: {
            commitment: 'commitment',
            size: 2,
          },
          previousBlockHash: 'burrito',
          randomness: '1',
          sequence: 2,
          target: 'icecream',
          timestamp: 200000,
          work: '123',
          hash: 'ramen',
        },
        transactions: [],
      }
      await peerNetwork['handleGossipMessage'](
        peer,
        new NewBlockMessage(block, Buffer.alloc(16, 'nonce')),
      )

      expect(peerNetwork['node']['syncer'].addNewBlock).toHaveBeenCalledWith(peer, block)
    })

    describe('when the worker pool is saturated', () => {
      it('does not accept or sync transactions', async () => {
        const node = mockNode()
        const peerNetwork = new PeerNetwork({
          identity: mockPrivateIdentity('local'),
          agent: 'sdk/1/cli',
          webSocket: ws,
          node,
          chain: {
            ...mockChain(),
            verifier: {
              verifyNewTransaction: jest.fn(),
            },
          },
          strategy: mockStrategy(),
          hostsStore: mockHostsStore(),
        })

        const { accounts, memPool, workerPool } = node
        jest.spyOn(workerPool, 'saturated').mockImplementationOnce(() => false)
        const acceptTransaction = jest.spyOn(memPool, 'acceptTransaction')
        const syncTransaction = jest.spyOn(accounts, 'syncTransaction')

        await peerNetwork['onNewTransaction']({
          peerIdentity: '',
          message: new NewTransactionMessage(Buffer.from(''), Buffer.alloc(16, 'nonce')),
        })

        expect(acceptTransaction).not.toHaveBeenCalled()
        expect(syncTransaction).not.toHaveBeenCalled()
      })
    })

    describe('when the node is syncing', () => {
      it('does not accept or sync transactions', async () => {
        const chain = {
          ...mockChain(),
          synced: false,
          verifier: {
            verifyNewTransaction: jest.fn(),
          },
        }
        const node = {
          ...mockNode(),
          chain,
        }
        const peerNetwork = new PeerNetwork({
          identity: mockPrivateIdentity('local'),
          agent: 'sdk/1/cli',
          webSocket: ws,
          node,
          chain,
          strategy: mockStrategy(),
          hostsStore: mockHostsStore(),
        })

        const { accounts, memPool } = node
        const acceptTransaction = jest.spyOn(memPool, 'acceptTransaction')
        const syncTransaction = jest.spyOn(accounts, 'syncTransaction')

        await peerNetwork['onNewTransaction']({
          peerIdentity: '',
          message: new NewTransactionMessage(Buffer.from(''), Buffer.alloc(16, 'nonce')),
        })

        expect(acceptTransaction).not.toHaveBeenCalled()
        expect(syncTransaction).not.toHaveBeenCalled()
      })
    })

    describe('when the worker pool is not saturated', () => {
      it('verifies transactions', async () => {
        const chain = {
          ...mockChain(),
          verifier: {
            verifyNewTransaction: jest.fn(),
          },
        }
        const node = {
          ...mockNode(),
          chain,
        }

        const peerNetwork = new PeerNetwork({
          identity: mockPrivateIdentity('local'),
          agent: 'sdk/1/cli',
          webSocket: ws,
          node,
          chain,
          strategy: mockStrategy(),
          hostsStore: mockHostsStore(),
        })

        // Spy on new transactions
        const verifyNewTransactionSpy = jest.spyOn(
          peerNetwork['chain']['verifier'],
          'verifyNewTransaction',
        )

        await peerNetwork['onNewTransaction']({
          peerIdentity: '',
          message: new NewTransactionMessage(Buffer.from(''), Buffer.alloc(16, 'nonce')),
        })

        expect(verifyNewTransactionSpy).toHaveBeenCalled()
      })
    })
  })

  describe('when enable syncing is false', () => {
    it('does not handle blocks', async () => {
      const peerNetwork = new PeerNetwork({
        identity: mockPrivateIdentity('local'),
        agent: 'sdk/1/cli',
        webSocket: ws,
        node: mockNode(),
        chain: mockChain(),
        strategy: mockStrategy(),
        hostsStore: mockHostsStore(),
        enableSyncing: false,
      })
      const { peer } = getConnectedPeer(peerNetwork.peerManager)
      const block = {
        header: {
          graffiti: 'chipotle',
          minersFee: '0',
          noteCommitment: {
            commitment: Buffer.from('commitment'),
            size: 1,
          },
          nullifierCommitment: {
            commitment: 'commitment',
            size: 2,
          },
          previousBlockHash: 'burrito',
          randomness: '1',
          sequence: 2,
          target: 'icecream',
          timestamp: 200000,
          work: '123',
          hash: 'ramen',
        },
        transactions: [],
      }

      const peerIdentity = peer.getIdentityOrThrow()
      const gossip = await peerNetwork['onNewBlock']({
        peerIdentity,
        message: new NewBlockMessage(block, Buffer.alloc(16, 'nonce')),
      })
      expect(gossip).toBe(false)
      expect(peerNetwork['node']['syncer'].addNewBlock).not.toHaveBeenCalled()
    })

    it('does not handle transactions', async () => {
      const peerNetwork = new PeerNetwork({
        identity: mockPrivateIdentity('local'),
        agent: 'sdk/1/cli',
        webSocket: ws,
        node: mockNode(),
        chain: mockChain(),
        strategy: mockStrategy(),
        hostsStore: mockHostsStore(),
        enableSyncing: false,
      })
      const { peer } = getConnectedPeer(peerNetwork.peerManager)

      const peerIdentity = peer.getIdentityOrThrow()
      const gossip = await peerNetwork['onNewTransaction']({
        peerIdentity,
        message: new NewTransactionMessage(Buffer.from(''), Buffer.alloc(16, 'nonce')),
      })
      expect(gossip).toBe(false)
      expect(peerNetwork['chain']['verifier'].verifyNewTransaction).not.toHaveBeenCalled()
    })
  })
})
