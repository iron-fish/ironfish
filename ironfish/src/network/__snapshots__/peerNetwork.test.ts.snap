// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registers a handler 1`] = `
Map {
  "NewBlock" => "gossip",
  "NewTransaction" => "gossip",
  "GetBlockHashes" => "directRPC",
  "GetBlocks" => "directRPC",
  "hello" => "gossip",
}
`;
